securesystemslib/spx_keys.py

```diff
@@ -0,0 +1,222 @@
+
+def generate_public_and_private():
+  """
+  <Purpose>
+    Generate a pair of spx public and private keys with pyspx.  The public
+    and private keys returned conform to 'SPX_PUBLIC_BYTES_SCHEMA' and
+    'SPX_PRIVATE_BYTES_SCHEMA', respectively.
+
+    An spx seed key is a random 128-byte string. Public keys are 64 bytes.
+
+    >>> public, private = generate_public_and_private()
+    >>> SPX_PUBLIC_BYTES_SCHEMA.matches(public)
+    True
+    >>> SPX_PRIVATE_BYTES_SCHEMA.matches(private)
+    True
+
+  <Arguments>
+    None.
+
+  <Exceptions>
+    NotImplementedError, if a randomness source is not found by 'os.urandom'.
+
+  <Side Effects>
+    The spx keys are generated by first creating a random seed
+    with os.urandom() and then calling pyspx's pyspx.signing.SigningKey().
+
+  <Returns>
+    A (public, private) tuple that conform to
+    'SPX_PUBLIC_BYTES_SCHEMA' and
+    'SPX_PRIVATE_BYTES_SCHEMA', respectively.
+  """
+
+  # Generate spx's seed key by calling os.urandom().  The random bytes
+  # returned should be suitable for cryptographic use and is OS-specific.
+  # Raise 'NotImplementedError' if a randomness source is not found.
+  seed = os.urandom(pyspx.crypto_sign_SEEDBYTES)
+
+  # Generate the public key.  pyspx performs the actual key generation.
+  public, private = pyspx.generate_keypair(seed)
+
+  return public, private
+
+
+
+
+
+def create_signature(private_key, data, scheme):
+  """
+  <Purpose>
+    Return a (signature, scheme) tuple, where the signature scheme is 'spx'
+    and is always generated by pyspx.  The signature returned
+    conforms to 'SPX_SIG_BYTES_SCHEMA'.
+
+    >>> public, private = generate_public_and_private()
+    >>> data = b'The quick brown fox jumps over the lazy dog'
+    >>> scheme = 'spx'
+    >>> signature, scheme = \
+        create_signature(private, data, scheme)
+    >>> SPX_SIG_BYTES_SCHEMA.matches(signature)
+    True
+    >>> scheme == 'spx'
+    True
+
+  <Arguments>
+    private:
+      The spx private key, a simple byte string
+
+    data:
+      Data object used by create_signature() to generate the signature.
+
+    scheme:
+      The signature scheme used to generate the signature.
+
+  <Exceptions>
+    securesystemslib.exceptions.FormatError, if the arguments are improperly
+    formatted.
+
+    securesystemslib.exceptions.CryptoError, if a signature cannot be created.
+
+  <Side Effects>
+    spx.signing.SigningKey.sign() called to generate the actual signature.
+
+  <Returns>
+    A signature dictionary conformant to
+    'securesystemslib.format.SIGNATURE_SCHEMA'.
+
+  """
+  # Validate arguments
+  SPX_PRIVATE_BYTES_SCHEMA.check_match(private_key)
+  securesystemslib.formats.SPX_SIG_SCHEMA.check_match(scheme)
+
+  try:
+    signature = pyspx.sign(data, private_key)
+
+  except (ValueError, TypeError) as e:
+    raise securesystemslib.exceptions.CryptoError('An "spx" signature'
+        ' could not be created with pyspx: ' + str(e))
+
+  return signature, scheme
```

Comment:
Well spotted!

Reply:
Well spotted indeed. ;) I briefly considered removing it but left it for consistency with the rest of the `create_signature` functions, and to keep the diff small.

Besides, it kinda does make sense if `scheme` becomes an optional argument with a default, as is the case in `ecdsa_keys.create_signature`. So that, if the caller does not pass a scheme, s/he learns about the default by looking at the return value.

*OTOH and in contrast to being consistent and keeping the diff small, I did remove the `public_key` argument from `spx_keys.create_signature`, because it makes no sense (see 
a67a6b3).*

If you're not opposed, I'd just leave it.

Reply:
That makes sense, it seems fine as is.